course/migrations/0048_delete_duplicate_enrollments.py

```diff
@@ -29,8 +29,8 @@ def delete_duplicates(apps, schema_editor):
 class Migration(migrations.Migration):
 
     dependencies = [
-        ('userprofile', '0003_auto_20160728_1139'),
-        ('course', '0046_coursemodule_reading_opening_time'),
+        ('userprofile', '0004_auto_20200721_1422'),
```

Comment:
Did you test the case in which you only change this userprofile dependency without swapping the order of the two course migrations? If you didn't, there's no need to spend more time on this. I just think this migration error is really strange and I don't see why we have to change the order of the course migrations. The original order matched the order of development and the migrations were generated with Django as usual. The original userprofile dependency is what it is also because it matches the situation when the feature was developed. Anyway, it looks like we have to change migrations afterwards and it should be fine in production too.

Reply:
I just did a quick test with the following changes:
```diff
diff --git a/course/migrations/0047_delete_duplicate_enrollments.py b/course/migrations/0047_delete_duplicate_enrollments.py
index 1d30bf63..73913e84 100644
--- a/course/migrations/0047_delete_duplicate_enrollments.py
+++ b/course/migrations/0047_delete_duplicate_enrollments.py
@@ -29,7 +29,7 @@ def delete_duplicates(apps, schema_editor):
 class Migration(migrations.Migration):

     dependencies = [
-        ('userprofile', '0003_auto_20160728_1139'),
+        ('userprofile', '0004_auto_20200721_1422'),
         ('course', '0046_coursemodule_reading_opening_time'),
     ]
```
and hand a similar error message as before:
```log
Operations to perform:
  Apply all migrations: admin, apps, auth, authtoken, contenttypes, course, deviations, diploma, exercise, external_services, inheritance, news, notification, sessions, social_django, threshold, userprofile
Running migrations:
  Applying userprofile.0004_auto_20200721_1422... OK
  Applying course.0047_delete_duplicate_enrollments...Traceback (most recent call last):
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.UndefinedColumn: column course_enrollment.language does not exist
LINE 1: ...tance_id", "course_enrollment"."user_profile_id", "course_en...
                                                             ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/srv/aplus/a-plus/manage.py", line 18, in <module>
    execute_from_command_line(sys.argv)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 375, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/core/management/base.py", line 323, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/core/management/base.py", line 364, in execute
    output = self.handle(*args, **options)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/core/management/base.py", line 83, in wrapped
    res = handle_func(*args, **kwargs)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/core/management/commands/migrate.py", line 232, in handle
    post_migrate_state = executor.migrate(
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/migrations/executor.py", line 117, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/migrations/executor.py", line 245, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/migrations/migration.py", line 124, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/migrations/operations/special.py", line 190, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/srv/aplus/a-plus/course/migrations/0047_delete_duplicate_enrollments.py", line 22, in delete_duplicates
    Enrollment.objects
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/models/query.py", line 710, in delete
    collector.collect(del_query)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/models/deletion.py", line 191, in collect
    new_objs = self.add(objs, source, nullable,
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/models/deletion.py", line 88, in add
    if not objs:
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/models/query.py", line 278, in __bool__
    self._fetch_all()
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/models/query.py", line 1242, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/models/query.py", line 55, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1142, in execute_sql
    cursor.execute(sql, params)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 76, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/utils.py", line 89, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/srv/aplus/venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: column course_enrollment.language does not exist
LINE 1: ...tance_id", "course_enrollment"."user_profile_id", "course_en...
                                                             ^
```
The mention about missing column `course_enrollment.language` lead me to believe that the migrations are ordered the wrong way around. Didn't spend much time testing other options since the changes in the initial commit seemed to work.